# Cross Validation Techniques

Importing Libraries

In [2]:
import pandas as pd
import numpy as np

Import Data

In [19]:
data = pd.read_csv(r"C:\Users\ABC\Desktop\data scientist\Machine L Projects\Classification\Cross Validation Techniques\file\cancer_dataset.csv")

In [20]:
data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [21]:
x = data.iloc[:,2:]
y = data.iloc[:,1]

In [22]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   radius_mean              569 non-null    float64
 1   texture_mean             569 non-null    float64
 2   perimeter_mean           569 non-null    float64
 3   area_mean                569 non-null    float64
 4   smoothness_mean          569 non-null    float64
 5   compactness_mean         569 non-null    float64
 6   concavity_mean           569 non-null    float64
 7   concave points_mean      569 non-null    float64
 8   symmetry_mean            569 non-null    float64
 9   fractal_dimension_mean   569 non-null    float64
 10  radius_se                569 non-null    float64
 11  texture_se               569 non-null    float64
 12  perimeter_se             569 non-null    float64
 13  area_se                  569 non-null    float64
 14  smoothness_se            5

In [23]:
x.isnull().sum()

radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed: 32                569
dtype: int64

Delect the Unnamed column from x

In [25]:
x.drop('Unnamed: 32',axis=1,inplace=True)

In [27]:
x.isnull().sum()

radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

In [28]:
y.isnull().sum()

0

# Finding the accuracy using Train Test split

Changing the random state the accuracy is changing here is the example

In [131]:
# importation od libraries od train test split
from sklearn.model_selection import train_test_split
## linear regression libraries
from sklearn.linear_model import LogisticRegression
for i in range(5):
    x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3,random_state=i)
    # chossing the model
    lrmodel = LogisticRegression()
    # fitting the model
    lrmodel.fit(x_train,y_train)
    # Finding Accuracy
    result = lrmodel.score(x_test,y_test)
    print(result)

0.9590643274853801
0.9298245614035088
0.9181286549707602
0.9298245614035088
0.9005847953216374


As you see the random state from 1 to 5 are different 
* random state = 1 accuracy is 0.9590643274853801
* random state = 2 accuracy is 0.9298245614035088
* random state = 3 accuracy is 0.9181286549707602
* random state = 4 accuracy is 0.9298245614035088
* random state = 5 accuracy is 0.9005847953216374

# Finding the accuracy using K Fold Cross Validation

importation of libraries of K Fold Cross Validation

In [64]:
from sklearn.model_selection import KFold, cross_val_score

In [95]:
k = KFold(10)

chossing the model

In [96]:
model = LogisticRegression()

In [112]:

cross = cross_val_score(model, x,y)

In [98]:
print(cross)

[0.92105263 0.93859649 0.95614035 0.94736842 0.95575221]


In [99]:
np.mean(cross)

0.9437820214252446

In [111]:
import warnings
warnings.filterwarnings('ignore')
model = LogisticRegression()
arr = [10,20,30,40,50,60,70,80]
for i in arr:
    k = KFold(i)
    cross = cross_val_score(model, x,y,cv=k)
    print(f"when kfold : {i}",np.mean(cross))

when kfold : 10 0.9367794486215537
when kfold : 20 0.9440270935960593
when kfold : 30 0.9438596491228068
when kfold : 40 0.9423809523809524
when kfold : 50 0.9443939393939393
when kfold : 60 0.9405555555555555
when kfold : 70 0.9446428571428571
when kfold : 80 0.9426339285714287


In [113]:
max([0.9367794486215537
,0.9440270935960593
,0.9438596491228068
,0.9423809523809524
,0.9443939393939393
, 0.9405555555555555
,0.9446428571428571
,0.9426339285714287])

0.9446428571428571

When the kFolf = 5

In [114]:
k = KFold(5)
model = LogisticRegression()
cross = cross_val_score(model, x,y,cv=k)
print(cross)
print(np.mean(cross))
# when kfold is 0.9437820214252446

[0.9122807  0.93859649 0.97368421 0.93859649 0.95575221]
0.9437820214252446


When kfold = 20

In [115]:
k = KFold(20)
model = LogisticRegression()
cross = cross_val_score(model, x,y,cv=k)
print(cross)
print(np.mean(cross))
#when kfold = 20 Accuracy is 0.9440270935960593

[0.93103448 0.82758621 0.96551724 0.93103448 0.96551724 0.93103448
 0.93103448 0.93103448 0.93103448 1.         0.96428571 1.
 0.96428571 0.96428571 0.92857143 0.92857143 0.89285714 0.96428571
 0.96428571 0.96428571]
0.9440270935960593


# Stratified K-fold Cross Validation

In [116]:
from sklearn.model_selection import StratifiedKFold

In [117]:
skmfold = StratifiedKFold(n_splits=5)

In [118]:
lr = LogisticRegression()

In [121]:
r = cross_val_score(lr,x,y,cv=skmfold)

In [123]:
r

array([0.92105263, 0.93859649, 0.95614035, 0.94736842, 0.95575221])

In [124]:
np.mean(r)

0.9437820214252446

# LOOCV - Leave one Out CV

In [125]:
from sklearn.model_selection import LeaveOneOut

In [126]:
lc = LeaveOneOut()

In [127]:
lmo = LogisticRegression()

In [128]:
p = cross_val_score(lmo,x,y,cv=lc)

In [130]:
np.mean(p)

0.945518453427065